### PR TITLE
chore: stream the attachments to the reporter

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: examples/todomvc/tests/integration.spec.ts [hunk 1]
```diff
@@ -22,23 +22,28 @@ test.describe('New Todo', () => {
...
     await newTodo.fill(TODO_ITEMS[1]);
     await newTodo.press('Enter');
+    await page.waitForTimeout(3000);
 
     // Make sure the list now has two todo items.
     await expect(page.getByTestId('todo-title')).toHaveText([
       TODO_ITEMS[0],
       TODO_ITEMS[1]
     ]);
+    await page.waitForTimeout(3000);
```
Comment (yury-s) [inline examples/todomvc/tests/integration.spec.ts:43]: revert changes in this file

File: packages/playwright-test/src/reporters/merge.ts [hunk 2]
```diff
@@ -39,14 +39,11 @@ export async function createMergedReport(config: FullConfigInternal, dir: string
 
 function patchAttachmentPaths(events: JsonEvent[], resourceDir: string) {
   for (const event of events) {
-    if (event.method !== 'onTestEnd')
+    if (event.method !== 'onAttachment')
```
Comment (yury-s) [inline packages/playwright-test/src/reporters/merge.ts:42]: You also need to patch testId for  onAttachment in ProjectNamePatcher below.

File: packages/playwright-test/src/util.ts [hunk 3]
```diff
@@ -261,7 +261,7 @@ export function getPackageJsonPath(folderPath: string): string {
   return result;
 }
 
-export async function normalizeAndSaveAttachment(outputPath: string, name: string, options: { path?: string, body?: string | Buffer, contentType?: string } = {}): Promise<{ name: string; path?: string | undefined; body?: Buffer | undefined; contentType: string; }>  {
+export async function normalizeAndSaveAttachment(outputPath: string, name: string, options: { path?: string, body?: string | Buffer, contentType?: string } = {}) {
```
Comment (yury-s) [inline packages/playwright-test/src/util.ts:264]: why did this change?

File: packages/playwright-test/src/worker/testInfo.ts [hunk 4]
```diff
@@ -332,15 +343,37 @@ export class TestInfoImpl implements TestInfo {
...
+  }
+
+  private _attach(name: string, attachment: TestInfo['attachments'][0]) {
     const step = this._addStep({
       title: `attach "${name}"`,
       category: 'attach',
       wallTime: Date.now(),
     });
-    this.attachments.push(await normalizeAndSaveAttachment(this.outputPath(), name, options));
+    this._attachmentsPush(attachment);
```
Comment (yury-s) [inline packages/playwright-test/src/worker/testInfo.ts:355]: Replace line 355-362 with `this._attachWithoutStep(attachment);`

File: packages/playwright-test/src/worker/testInfo.ts [hunk 5]
```diff
@@ -157,6 +161,13 @@ export class TestInfoImpl implements TestInfo {
       const relativeTestFilePath = path.relative(this.project.testDir, test._requireFile);
       return path.join(this.project.snapshotDir, relativeTestFilePath + '-snapshots');
     })();
+
+    this._attachmentsPush = this.attachments.push.bind(this.attachments);
+    this.attachments.push = (...attachments: TestInfo['attachments']) => {
```
Comment (yury-s) [inline packages/playwright-test/src/worker/testInfo.ts:166]: do we want to handle removal of an attachment from the list?

**Review by yury-s — APPROVED** [review]

**Review by yury-s — APPROVED** [review]

File: packages/playwright-test/src/util.ts (hunk 3 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/util.ts:264]: Long type, while the inference is doing a good job.

File: packages/playwright-test/src/worker/testInfo.ts (hunk 5 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/worker/testInfo.ts:166]: I don't think so!

**Pull request closed without merging**